frame/lending/src/repay_borrow.rs

```diff
@@ -0,0 +1,136 @@
+use composable_traits::defi::DeFiComposableConfig;
+use frame_support::{
+	traits::fungibles::{Inspect, Mutate, MutateHold, Transfer},
+	DebugNoBound,
+};
+use sp_runtime::DispatchError;
+
+use crate::Config;
+
+/// Repay `amount` of `beneficiary_account`'s `borrow_asset` debt principal.
+///
+/// Release given `amount` of `debt_token` from `beneficiary_account`, transfer `amount` from
+/// `payer_account` to `market_account`, and then burn `amount` of `debt_token` from
+/// `beneficiary_account`.
+#[derive(Copy, Clone, DebugNoBound, PartialEq, Eq)]
+pub(crate) struct RepayPrincipal<'a, T: Config> {
+	/// The borrowed asset being repaid.
+	pub(crate) borrow_asset: <T as DeFiComposableConfig>::MayBeAssetId,
+
+	/// The debt token to burn from `beneficiary_account`.
+	pub(crate) debt_token: <T as DeFiComposableConfig>::MayBeAssetId,
+
+	/// The account repaying `beneficiary_account`'s debt.
+	pub(crate) payer_account: &'a T::AccountId,
+
+	/// The market account that will be repaid.
+	pub(crate) market_account: &'a T::AccountId,
+
+	/// The account that took out the borrow and who's debt is being repaid, i.e. the
+	/// beneficiary.
+	pub(crate) beneficiary_account: &'a T::AccountId,
+
+	/// The amount of `beneficiary_account`'s debt to be repaid by `payer_account`.
+	///
+	/// NOTE: This is assumed to be `<=` the total principal amount.
+	pub(crate) amount_of_debt_to_repay: <T as DeFiComposableConfig>::Balance,
+
+	// Whether or not to keep `from_account` alive.
+	pub(crate) keep_alive: bool,
+}
+
+impl<'a, T: Config> RepayPrincipal<'a, T> {
+	/// See the type level docs for [`RepayPrincipal`].
+	pub(crate) fn run(self) -> Result<(), DispatchError> {
```

Comment:
this is trait extending MultiCurrency, not trait on Self (see reserve exchange in acution PR)